docs/architecture/adr-020-protobuf-transaction-encoding.md

```diff
@@ -24,76 +25,186 @@ With this in mind, we will tackle the migration path via two main areas, txs and
 querying. However, this ADR solely focuses on transactions. Querying should be
 addressed in a future ADR, but it should build off of these proposals.
 
+Based on detailed discussions ([\#6030](https://github.com/cosmos/cosmos-sdk/issues/6030)
+and [\#6078](https://github.com/cosmos/cosmos-sdk/issues/6078)), the original 
+design for transactions was changed substantially from an `oneof` /JSON-signing
+approach to the approach described below.
+
 ## Decision
 
 ### Transactions
 
-Since the messages that an application is known and allowed to handle are specific
-to the application itself, so must the transactions be specific to the application
-itself. Similar to how we described in [ADR 019](./adr-019-protobuf-state-encoding.md),
-the concrete types will be defined at the application level via Protobuf `oneof`.
+Since interface values are encoded with `google.protobuf.Any` in state (see [ADR 019](adr-019-protobuf-state-encoding.md)),
+ `sdk.Msg`s are encoding with `Any` in transactions.
+ 
+ One of the main goals of using `Any` to encode interface values is to have a
+ core set of types which is reused by apps so that
+ clients can safely be compatible with as many chains as possible.
+ 
+ It is one of the goals of this specification to provide a flexible cross-chain transaction
+ format that can serve a wide variety of use cases without breaking client 
+ compatibility.
+
+In order to facilitate signing, transactions are separated into `TxBody`,
+which will be re-used by `SignDoc` below, and `signatures`:
+
+```proto
+// types/types.proto
+package cosmos_sdk.v1;
+
+message Tx {
+    TxBody body = 1;
+    repeated bytes signatures = 2;
+}
+
+message TxBody {
+    repeated google.protobuf.Any messages = 1;
+    repeated SignerInfo signer_info = 2;
+    Fee fee = 3;
+    string memo = 4;
+    int64 timeout_height = 5;
+    repeated google.protobuf.Any extension_options = 6;
+}
+
+message SignerInfo {
```

Comment:
I'd prefer to call this `Signature`.

Reply:
But it doesn't actually include the `Signature` at all, it just includes info about the signer that is making making the signature.

Reply:
ahhh, right yes. That's a bit confusing. Why is someone expected to sign over their pubkey? The signatures of a `Tx` should have the signature and pubkey, no? The signing doc only includes the mode along with everything else needed.

Reply:
I guess I'm trying to leave as little out of the signdoc as possible based on the precautionary principle. What some pubkey type stores extra info (like allowed signing modes as Zaki proposed)? Then l should be signing the pubkey because there could actually be two "keys" with the same signature but different security behavior.

I would also ask the counter question, what is the harm in signing the pubkeys?

Reply:
There is no harm at all. This is OK as long as it's documented 👍